src/LowerParallelTasks.cpp

```diff
@@ -0,0 +1,508 @@
+#include "LowerParallelTasks.h"
+
+#include <string>
+
+#include "Argument.h"
+#include "Closure.h"
+#include "DebugArguments.h"
+#include "ExprUsesVar.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "Module.h"
+#include "Param.h"
+
+namespace Halide {
+namespace Internal {
+
+// TODO(zalman): Find a better place for this code to live.
+LoweredFunc GenerateClosureIR(const std::string &name, const Closure &closure,
+                              std::vector<LoweredArgument> &args, int closure_arg_index,
+                              const Stmt &body, const Target &t) {
+    // Figure out if user_context has to be dealt with here.
+    std::string closure_arg_name = unique_name("closure_arg");
+    args[closure_arg_index] = LoweredArgument(closure_arg_name, Argument::Kind::InputScalar,
+                                              type_of<uint8_t *>(), 0, ArgumentEstimates());
+    Expr closure_arg = Variable::make(type_of<void *>(), closure_arg_name);
+
+    Stmt wrapped_body = body;
+    std::vector<Expr> type_args(closure.vars.size() + closure.buffers.size() * 2 + 2);
+    std::string closure_type_name = unique_name("closure_struct_type");
+    Expr struct_type = Variable::make(Handle(), closure_type_name);
+    type_args[0] = StringImm::make(closure_type_name);
+    type_args[1] = 1;
+
+    int struct_index = 0;
+    for (const auto &v : closure.vars) {
+        type_args[struct_index + 2] = make_zero(v.second);
+        wrapped_body = LetStmt::make(v.first,
+                                     Call::make(v.second, Call::load_struct_member,
+                                                {closure_arg, struct_type, make_const(UInt(32), struct_index)},
+                                                Call::Intrinsic),
+                                     wrapped_body);
+        struct_index++;
+    }
+    for (const auto &b : closure.buffers) {
+        type_args[struct_index + 2] = make_zero(type_of<void *>());
+        type_args[struct_index + 3] = make_zero(type_of<halide_buffer_t *>());
+        wrapped_body = LetStmt::make(b.first,
+                                     Call::make(type_of<void *>(), Call::load_struct_member,
+                                                {closure_arg, struct_type, make_const(UInt(32), struct_index)},
+                                                Call::PureIntrinsic),
```

Comment:
This call is `PureIntrinsic` but the other two are `Intrinsic`. Is this intentional? If so, we should comment why. If not, we should fix it.

Reply:
Resolved?